src/ray/core_worker/core_worker.cc

```diff
@@ -459,15 +459,44 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
       {{"worker_id", worker_id.Hex()}});
 }
 
+CoreWorker::~CoreWorker() { RAY_LOG(INFO) << "Core worker is destructed"; }
+
 void CoreWorker::Shutdown() {
-  io_service_.stop();
+  RAY_LOG(INFO) << "Shutting down a core worker.";
+  is_shutdown_ = true;
   if (options_.worker_type == WorkerType::WORKER) {
+    // Running in a main thread.
     direct_task_receiver_->Stop();
     task_execution_service_.stop();
   }
   if (options_.on_worker_shutdown) {
+    // Running in a main thread.
     options_.on_worker_shutdown(GetWorkerID());
   }
+
+  io_service_.stop();
+  if (gcs_client_) {
+    // TODO(sang): This causes a thread-safety bug because
```

Comment:
By reverting the previous fix, does it make the tests pass?

Reply:
Seems like it does. I actually found there was one edge case with HA Ray upon GCS shutdown with previous fix. I have a new fix in my mind. We can discuss in details in the next PR